zip-0221.rst

```diff
@@ -0,0 +1,585 @@
+
+Personalized BLAKE2b-256 was selected to match existing ZCash conventions. Adding the consensus branch ID to the hash personalization string ensures that valid nodes from one branch cannot be used to make false statements about parallel consensus branches.
+
+FlyClient Requirements and Recommendations
+===============================================
+These commitments enable FlyClient in the variable-difficulty model. Specifically, they allow light clients to reason about application of the difficulty adjustment algorithm over a range of blocks. They were chosen via discussion with an author of the FlyClient paper.
+
+- ``nEarliestTimestamp``
+- ``nLatestTimestamp``
+- ``nEarliestTarget``
+- ``nLatestTarget``
+- ``nEarliestHeight``
+- ``nLatestHeight``
+- ``nSubTreeTotalWork``
+
+Non-FlyClient Commitments
+==========================
+
+Additional metadata commitments were chosen primarily to improve light client security guarantees. We specified commitments where we could see an obvious security benefit, but there may be other useful metadata that we missed. We're interested in feedback and suggestions from the implementers of the current light client.
+
+We considered adding a commitment to the nullifier vector at each block. We would appreciate comments from light client teams on the utility of this commitment, as well as the proper serialization and commitment format for the nullifier vector.
+
+- ``hashEarliestSaplingRoot``
+
+  * Committing to the earliest Sapling root of a range of blocks allows light clients to check the consistency of treestate transitions over a range of blocks, without recalculating the root from genesis.
+
+- ``hashLatestSaplingRoot``
+
+  * This commitment serves the same purpose as ``hashFinalSaplingRoot`` in current Sapling semantics.
+  * However, because the MMR tree commits to blocks ``B_x ... B_(n-1)``, the latest commitment will descrie the final treestate of the previous block, rather than the current block.
+  * Concretely: block 500 currently commits to the final treestate of block 500 in its header. With this ZIP, block 500 will commit to all roots up to block 499, but not the final root of block 500.
+  * We feel this is an acceptable tradeoff. Using the most recent treestate as a transaction anchor is already unsafe in reorgs. Clients should never use the most recent treestate to generate transactions, so it is acceptable to delay commitment by one block.
+
+- ``nShieldedTxCount``
+
+  * By committing to the number of shielded transactions in blocks (and ranges of blocks), a light client may reliably learn whether a malicious server is witholding any shielded transactions.
```

Comment:
This only holds if the light client is expecting to see all shielded transactions. Sapling-based light clients only download Sapling-related transactions (for bandwidth efficiency, and because there is no shared anonymity set across Sprout and Sapling), and thus would not benefit from this (as the server could claim that all shielded transactions within the block were non-Sapling). It might be better to have separate commitments for Sprout-containing transactions and Sapling-containing transactions; this also then leads to the earlier question of how to extend the MMR node structure to handle future shielded pools.